File: leetcode/problems/language-specific/go/78.subsets.go
/*
 * @lc app=leetcode id=78 lang=golang
 *
 * [78] Subsets
 */

// @lc code=start
func subsets(nums []int) [][]int {
	if len(nums) <= 0 {
		return [][]int{}
	}

	res := [][]int{}
	subsetsHelper(nums, 0, []int{}, &res)
	return res
}

func subsetsHelper(nums []int, start int, subset []int, res *[][]int) {
	*res = append(*res, append([]int{}, subset...))

	for i := start; i < len(nums); i++ {
		subset = append(subset, nums[i])
		subsetsHelper(nums, i+1, subset, res)
		subset = subset[:len(subset)-1]
	}
}

// @lc code=end

File: leetcode/problems/language-specific/python/78.subsets.py
#
# @lc app=leetcode id=78 lang=python3
#
# [78] Subsets
#

# @lc code=start

# * Backtracking Solution | O(n*(2^n)) Time | O(n) Space
# * n -> The length of nums array


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []

        def subsetsHelper(start_idx, subset):
            res.append(subset[:])

            for i in range(start_idx, len(nums)):
                subset.append(nums[i])
                subsetsHelper(i + 1, subset)
                subset.pop()

        subsetsHelper(0, [])
        return res


# @lc code=end
